### DISCUSSION TITLE
Apply Primitive Transformers (like Catch22) to Rolling Windows

### DISCUSSION THREAD

**Discussion by LMolr** [alan-turing-institute/sktime#7248] [Q&A] [answered]
Hi,
is it possible to apply a "Primitive"-output Transformer (like Catch22) to a time-series in a *rolling window* fashion?

Taking Catch22 as example: given a series, it outputs 22 scalars.
I'd like to build a Transformer which, given a series, outputs 22 series, each series being a _Catch22_ scalar computed on a past rolling time window.

Is it achievable with the current SKTime building blocks?

Thanks!

**Comment by fkiraly** [comment, marked as answer]
Yes, via the `SplitterSummarizer`, this allows to vary the splitting schema and the transformer independently:
https://www.sktime.net/en/latest/api_reference/auto_generated/sktime.transformations.series.summarize.SplitterSummarizer.html

Certain simple summarizations are more efficiently implemented in the `WindowSummarizer`, but that does not allow to pick arbitrary transformations.

Finally, you can also use the `TemporianTransformer` to use the `temporian` feature extraction framework in `sktime`, or the `PandasTransformAdaptor` to use `pandas` methods like `rolling`, though both rely on external frameworks and also do not allow to use arbitrary `sktime` transformations.

**Comment by fkiraly** [comment]
Although, I notice that `SplitterSummarizer` does not exactly fit your specification.

Do you mean, you want 22 *variables* or 22 individual *series*? `SplitterSummarizer` will produce *a single multivariate series with 22 variables*, corresponding to the 22 Catch-22 features.

If you want to create 22 series, I am not sure how to do this currently. This could be achieved with a composition that turns variables into series, e.g., moves col index to a new row index level in pandas, using `pivot` or similar. Possibly the `PandasTransformAdaptor` can be used here.

Could you explain which of the two you want? If the latter, what would be a use case?

**Reply by LMolr** [reply]
If I understand correctly, I need 22 variables.

Assuming i have a `y` series of length 5:

| t   | y   |
| --- | --- |
| t0  | y0  |
| ... | ... |
| t4  | y4  |

The output I expect with window_size=3 is a DataFrame like this:

| t   | catch_1 | catch_2 | ... | catch_22 | Note                          |
| --- | ------- | ------- | --- | -------- | ----------------------------- |
| t3  | ... | ...     | ... | ...      | _Computed as `catch22(y0, y1, y2)`_ |
| t4  | ...     | ...     | ... | ...      | _Computed as `catch22(y1, y2, y3)`_ |

_In this example, i dropped the first rows of the output because there is not enough history to respect `window_
length=3`_

I hope this clears a bit my question.
Thanks!

**Comment by LMolr** [comment]
@fkiraly , it seems that your first suggestion fits my case.

For instance, it seems that I can successfully combine two Transformers and apply them to rolling windows.

```python
def new_xf_windowed_summ(window_length):

    xf_prims = FeatureUnion([
        Catch22(catch24=True, col_names="short_str_feat"),
        HurstExponentTransformer(min_lag=2, max_lag=10),
    ])

    xf_splitter = SlidingWindowSplitter(
        window_length=window_length,
        # Not sure about the following ...
        step_length=1,
        fh=[0],
    )

    xf_summ = SplitterSummarizer(xf_prims, xf_splitter, index='last')

    return xf_summ

window_length = 10

xf_windowed_summ = new_xf_windowed_summ(window_length=window_length)
y_trans = xf_windowed_summ.fit_transform(y)
```

I am just not quite sure about some SlidingWindow arguments...

Do you have any comments (caveats etc.) about the approach?

Thank you so much!

**Reply by fkiraly** [reply]
Yes, indeed it seems this is what you were looking for.

I cannot think of any caveat, maybe here´s one: the default setting of `SplitterSummarizer` uses only the train window from the splitter. Arguments determining the test window, e.g., `fh` are ignored. Alternative uses are controlled via `fit_on` and `transform_on`.
